docs/RFCS/20190318_error_handling.md

```diff
@@ -0,0 +1,1882 @@
+type, as happens currently in CockroachDB when a non-`roachpb` error
+is transferred through the network. When wire representations *are*
+available, the method is generally reliable; however, if errors are
+implemented as a chain of causes, care should be taken to perform the
+test on all the intermediate levels. See instances in section
+[Suspicious assertions on the error
+type](#Suspicious-assertions-on-the-error-type).
+
+Method 3 is generally reliable although the predicates in the standard
+library obviously do not know about any additional custom types. Also,
+the implementation of the predicate method can be cumbersome if one
+must test errors from multiple packages (dependency cycles). Also, the
+method loses its reliability if the predicate itself relies on one of
+the other methods in a way that's unreliable. For example, the current
+predicates in CockroachDB's `sqlbase` package are defective in this
+way.
+
+Method 4 is the most problematic and unfortunately has been used
+pervasively inside CockroachDB. It has several sub-problems, detailed
+in the following sub-sections. See also the section [Suspicious
+comparisons of the error
+message](#Suspicious-comparisons-of-the-error-message) at the end for
+the list of suspicious cases in the code.
+
+### Vulnerability to embedded application strings
+
+The main problem with comparing an error message to some reference
+string is that the reference can appear in one of the application
+values embedded inside the error. For example, consider the error
+produced thus:
+
+```
+root@127.0.0.1:34312/defaultdb> select 'connection reset by peer'::int;
+pq: could not parse "connection reset by peer" as type int: strconv.ParseInt: parsing "connection reset by peer": invalid syntax
+```
+
+And the test in `pgwire/pgerror/errors.go` function
+`IsSQLRetryableError()` which merely checks whether the error contains
+the string `"connection reset by peer"`. This method would return
+`true` in that case, whereas the error is not retriable.
+
+This problem is in fact a form of *bad value injection* and creates a
+vector by which a remote user can misuse the internals of CockroachDB
+(a security vulnerability).
+
+**In general, string comparisons on error messages are vulnerable to
+injection and can even cause security problems.**
+
+### Improper/unsafe testing of intermediate causes
+
+When, say, a retry error happens while handling a duplicate insertion error,
+we want a quick way to determine the error is retryable.
+
+Testing whether the entire error message contains some reference
+string can work but is neither fast nor reliable (see previous section).
+
+In the (relatively uncommon case) of a constant string prefix, one can
+rely on exact matches to the first argument provided to
+`errors.Wrap`. In that case, the prefix provided as 2nd
+argument to `errors.Wrap()` can be extracted as follows:
+
+1. `s1 := err.Error()`
+2. `s2 := err.Cause().Error()`
+3. `prefix_message := s1[:len(s2)-len(s1)]`
```

Comment:
Oof, this may also break because who's to say that the error was created with `errors.Wrap`? Anyone can implement `causer`.